### PR TITLE
[`_unmask_unattended`] Refactor

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Follow up on #29318, if we keep the `AttentionMaskConverter` then let's put the checks in the `_unmask_unattended` since this function is bound to disappear once the bug is fixed in torch

@fxmarty let's make this easier to refactor later on, and more readable!

### CODE REVIEW

File: src/transformers/modeling_attn_mask_utils.py [hunk 1]
```diff
@@ -227,12 +229,24 @@ def _unmask_unattended(
...
-        return expanded_mask.mul(~torch.all(expanded_mask == min_dtype, dim=-1, keepdim=True))
+        if attention_mask is not None and attention_mask.device.type == "cuda":
+            if expanded_mask.dtype == torch.bool:
+                raise ValueError(
+                    "AttentionMaskConverter._unmask_unattended expects a float `expanded_mask`, got a BoolTensor."
+                )
+
+            # TODO: For dynamo, rather use a check on fullgraph=True once this is possible (https://github.com/pytorch/pytorch/pull/120400).
+            is_tracing = torch.jit.is_tracing() or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
+            is_tracing |= isinstance(input_tensor, torch.fx.Proxy)
```
Comment (fxmarty) [inline src/transformers/modeling_attn_mask_utils.py:241]: You can leave this as a one-liner. I believe in python `a or b` never evaluates b if a is already True.

Comment (fxmarty) [inline src/transformers/modeling_attn_mask_utils.py:241]: Can you pass `is_tracing` as an `Optional[bool]`? This variable is already computed in the `_prepare_xxx` methods. And check it here only if it is None.

File: src/transformers/modeling_attn_mask_utils.py [hunk 2]
```diff
@@ -188,6 +188,8 @@ def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int]
     @staticmethod
     def _unmask_unattended(
         expanded_mask: torch.FloatTensor,
+        attention_mask: torch.FloatTensor,
+        input_tensor: torch.FloatTensor,
```
Comment (fxmarty) [inline src/transformers/modeling_attn_mask_utils.py:192]: I don't think we need `input_tensor`, do we? The FX check should be doable on `attention_mask`

Comment (ArthurZucker) [reply src/transformers/modeling_attn_mask_utils.py:192]: but sometimes the attention mask is None even when tracing no?

Comment (fxmarty) [reply src/transformers/modeling_attn_mask_utils.py:192]: I meant `expanded_mask`. See the typing

Comment (ArthurZucker) [reply src/transformers/modeling_attn_mask_utils.py:192]: alright
